### ISSUE TITLE
6月底拟迁移到module的算法

### ISSUE THREAD

**Issue by JiaqingFu** [Oneflow-Inc/models#3]
# 6月底拟迁移的算法
下面这个表不方便更新，在[这个表](https://docs.qq.com/sheet/DQ3BodU9sSmNQRk9v?tab=BB08J2)更新

| model                                                        | 负责人                 | 缺失算子 | pr |
| ------------------------------------------------------------ | ---------------------- | ------------------------------------------------ |---- |
| [ResNet50](https://github.com/Oneflow-Inc/models/tree/main/resnet50) |         梁德澎(done)                   |                    |
|  [VGG19](https://github.com/Oneflow-Inc/models/tree/main/vgg)  |             徐之秋(done)                 |             |
|  [VGG16](https://github.com/Oneflow-Inc/models/tree/main/vgg) |               徐之秋(done)               |             |
|  [MobilenetV3](https://github.com/Oneflow-Inc/models/tree/main/mobilenetv3) |       张晓雨(done)                                |          |
|  [MobilenetV2](https://github.com/Oneflow-Inc/models/tree/main/mobilenetv2) |         赵露阳(done)                                |        |
|  [Shufflenet](https://github.com/Oneflow-Inc/models/tree/main/shufflenetv2) |            姚迟(done)                                |      |
|  [RNN](https://github.com/Oneflow-Inc/models/tree/main/rnn) |              徐之秋(done)                         |           |
|  [LSTM](https://github.com/Oneflow-Inc/models/tree/main/rnn) |             徐之秋(done)                           |     flow.tensor.view     |
|  [BERT](https://github.com/Oneflow-Inc/models/tree/main/bert-oneflow) |              李春游、曾畅、赵露阳、张晓雨 |          |
|  TextCNN |        杨羿、陆陈昊               |      flow.optimizer.PiecewiseConstantScheduler              |
| [Lenet](https://github.com/Oneflow-Inc/models/tree/main/quick_start_demo_lenet) |        张子涵             | 同上                  |
| face_seg |      魏日令、叶娇娇                 |nn.ConvTranspose2d，nn.BCEWithLogitsLoss，optim.lr_scheduler.CosineAnnealingLR|
| [FastNeuralStyle](https://github.com/Oneflow-Inc/models/tree/main/fast_neural_style) |        徐之秋、梁德澎(done)                       |               |
| SRGAN |       孙立剑、陈岱渊               |     loss依赖vgg19 ,以下三个算法缺失算子见comment           |   https://github.com/Oneflow-Inc/models/pull/56
| [PIX2PIX](https://github.com/Oneflow-Inc/models/tree/main/pix2pix)|       陈岱渊 、吴方闻 （done）          |                  | https://github.com/Oneflow-Inc/models/pull/30
| [DCGAN](https://github.com/Oneflow-Inc/models/tree/main/DCGAN) |      陈岱渊 、吴方闻 （done）           |                |https://github.com/Oneflow-Inc/models/pull/17  
| [ReID](https://github.com/Oneflow-Inc/models/tree/main/reid) |      吴翠玲、钟昊文 (done)             |                   |  https://github.com/Oneflow-Inc/models/pull/36
| TSN |       李玲、王一张             |   variance_scaling_initializer ，使用resnet50 作为 backbone                | 
| YOLOV3 |          王世杰                 |                   | 
| maskrcnn|          姜雪飞                 |                   | 
| restrack  |       张文广             |     依赖YOLOV3与ReID              | 
| [FlappyBird](https://github.com/Oneflow-Inc/models/tree/main/FlappyBird) |       陆陈昊、王一张(done)                    |      torch.manual_seed             | https://github.com/Oneflow-Inc/models/pull/39
| ProGAN |   李融  |   nn.ConvTranspose2d        |

[算法库链接](https://github.com/Oneflow-Inc/oneflow_vision_model)，建议对照pytorch代码迁移算法，而不是oneflow旧接口代码

**Comment by Ldpe2G** [comment]
# 模型验收标准第一版

复现好的模型，代码提交 [Oneflow-Inc/models](https://github.com/Oneflow-Inc/models) 仓库统一管理。

对于能找到Pytorch官方实现的模型，比如视觉分类网络，直接对齐torch实现然后把torch的预训练模型直接转成oneflow来用，需要提供 预训练模型、预测脚本和README，完整的训练代码可以后期再补充。

如果是从原来lazy实现转成eager的模型，则需要提供完整的训练脚本、预训练模型、预测脚本和README。

README 的内容和格式参考：[resnet50](https://github.com/Oneflow-Inc/models/blob/main/resnet50/README.md)，如果是生成类或者像素级任务，比如 GAN，风格化、分割和超分等等，则需要在 README.md 中提供可视化的结果。

预训练模型需要上传到阿里云，可以暂时先发给我统一上传。

预测或者训练python脚本需同时提供 bash 脚本，用于方便用户设置一些可变的命令行参数。比如预测脚本，可在bash脚本中下载预训练模型，尽量要让用户下载代码装好oneflow之后，直接运行 bash 脚本就能跑起来。比如 [resnet50/infer.sh](https://github.com/Oneflow-Inc/models/blob/main/resnet50/infer.sh)

**Comment by yayeoCddy** [comment]
所有的GAN相关训练都需要基于`torch.Tensor.detach` 这个算子(目前是@liyurui  在修)

图像生成/超分任务Pix2pix, DCGAN, SRGAN模型需要迁移以下算子：
|                          |                                           | 备注                                   |
|--------------------------|-------------------------------------------|----------------------------------------|
| torch.nn.ConvTranspose2d | flow.nn.conv2d_transpose                  | 不太清楚oneflow原本这两个反卷积的区别  |
|                          | flow.nn.torch_conv2d_transpose            |                                        |
|                          | flow.nn.compat_conv2d                     | 不太清楚和已搬运的flow.nn.conv2d的区别 |
|                          | flow.nn.bias_add                          | 动态图中是否还需要？                   |
|                          | flow.nn.sigmoid_cross_entropy_with_logits |                                        |
| torch.pad                | flow.pad                                  |                                        |
|                          | flow.constant_initializer                 | 好像已经搬运到flow.nn.init下           |
|                          | flow.zeros_initializer                    |                                        |
|                          | flow.ones_initializer                     |                                        |
|                          | flow.nn.max_pool2d                        | 已搬                                   |
|                          | flow.nn.avg_pool2d                        |                                        |
|                          | flow.math.reduce_mean                     |                                        |
|                          | flow.math.reduce_sum                      |                                        |
|                          | flow.math.squared_difference              | flow是否可以拼？有无搬运必要？         |
|                          | flow.slice                                | 已搬?                                  |

**Comment by LukeLuuu** [comment]
restrack模型需要迁移以下算子：
|                          |                                           | 备注                                   | 认领人 |
|--------------------------|-------------------------------------------|----------------------------------------|--------|
|  | flow.nn.bias_add()                  |   |        |
|                          | flow.layers.upsample_2d()            |                                        |        |
|                          | flow.yolo_detect()                     |  |        |
|                          | flow.math.logistic()                          |                    |        |
|                          | flow.yolo_box_diff() |                                        |        |
|                 | flow.math.top_k()                                  |                  已搬？                      |        |
|                          | flow.yolo_nms()                 |           |        |

**Comment by JiaojiaoYe1994** [comment]
|      | 所需算子                                  | 对齐目标                                                     | 目标算子                            | 备注                                                         |
| ---- | ----------------------------------------- | ------------------------------------------------------------ | ----------------------------------- | ------------------------------------------------------------ |
| 1    | flow.variance_scaling_initializer         |                                                              |                                     | 缺失                                                         |
| 2    | flow.regularizers.l2                      |                                                              |                                     | 缺失                                                         |
| 3    | flow.zeros_initializer                    |                                                              |                                     | 缺失                                                         |
| 4    | flow.get_variable                         |                                                              |                                     | 缺失                                                         |
| 5    | flow.nn.conv2d                            | nn.Conv2d                                                    | nn.Conv2d                           | 已搬运                                                       |
| 6    | flow.nn.conv2d_transpose                  | nn.ConvTranspose2d                                           | nn.conv2d_transpose                 | 已搬运                                                       |
| 7    | flow.nn.bias_add                          |                                                              |                                     | 缺失                                                         |
| 8    | flow.layers.batch_normalization           | nn.BatchNorm2d                                               | flow.nn.BatchNorm2d                 | 已搬运                                                       |
| 9    | flow.math.relu                            | nn.ReLU                                                      | flow.nn.relu                        | 已搬运                                                       |
| 10   | flow.nn.max_pool2d                        | nn.MaxPool2d                                                 | flow.nn.MaxPool2d                   | 已搬运                                                       |
| 11   | flow.nn.sigmoid_cross_entropy_with_logits | nn.BCEWithLogitsLoss                                         | nn.BCEWithLogitsLoss                | 叶娇娇nn.BCEWithLogitsLoss                                   |
| 12   | flow.math.reduce_mean                     |                                                              | flow.math.reduce_mean               | 缺失                                                         |
| 13   | flow.math.sigmoid                         | torch.nn.functional.sigmoid ，torch.sigmoid torch.nn.Sigmoid | oneflow.math.sigmoid                | torch.nn.Sigmoid 已经搬运，缺新后向                          |
| 14   | flow.math.log                             | torch.log                                                    | oneflow.math.log                    | 已搬运，黄振华 https://github.com/Oneflow-Inc/oneflow/pull/5006 |
| 15   | flow.FixedTensorDef                       |                                                              |                                     | 缺失                                                         |
| 16   | flow.float                                |                                                              |                                     | 缺失                                                         |
| 17   | flow.reshape                              |                                                              | oneflow.reshape                     | 已搬运                                                       |
| 18   | flow.optimizer.PiecewiseScalingScheduler  |                                                              | optimizer.PiecewiseScalingScheduler | 缺失                                                         |
| 19   | flow.optimizer.SGD                        | torch.optim.SGD                                              | optimizer.SGD                       | 已搬运                                                       |
| 20   | flow.env.log_dir                          |                                                              |                                     | 缺失                                                         |
| 21   | flow.train.CheckPoint                     |                                                              |                                     | 缺失                                                         |
| 22   | flow.function_config                      |                                                              |                                     | 缺失                                                         |
| 23   | flow.config.gpu_device_num                |                                                              |                                     |                                                              |

**Comment by puchapu** [comment]
ReID 算法所需算子
|pytorch                          |oneflow待搬运                                           | 备注                                   | 认领人 |
|--------------------------|-------------------------------------------|----------------------------------------|--------|
| torch.addmm |                | 通过 flow.multipy 和 flow.add 拼  |        |
|  torch.nn.functional.one_hot                        |  flow.one_hot      |                                        | 钟昊文       |
|  torch.nn.marginrankingloss                   |                    | 可以 python 端拼|     钟昊文   |
|  torch.max/min                        | flow.math.reduce_max/sum                        |                |        |
|            |  flow.tripletloss          |                      可以 python 端拼               |  钟昊文     |
|            |  flow.train.CheckPoint            |                                     |        |
|            |  flow.optimizer.PiecewiseScalingScheduler         |                                     |        |
|            | flow.optimizer.warmup.linear        |                                     |        |

**Comment by simonJJJ** [comment]
Mask R-CNN所需算子：
|pytorch                          |oneflow待搬运                                           | 备注                                   | 认领人 |
|--------------------------|-------------------------------------------|----------------------------------------|--------|
|torch.nn.ConvTranspose2d|flow.nn.ConvTranspose2d||张晓雨|
|torch.topk|flow.topk|||
|torch.split|flow.split|||
|torch.max|flow.max|||
|torch.min|flow.min|||
|torch.stack|flow.stack|||
|torch.sigmoid|flow.sigmoid|||
|torch.randperm|flow.randperm|||
|torch.zeros_like|flow.zeros_like|||
|roi_align(外部op)|roi_align|||
|nms(外部op)|nms|||

**Comment by doombeaker** [comment]
> sigmoid

这里的 topk、max、min 之江黄振华正在做。sigmoid 已经有了。zeros_like应该可以已有的用zeros替代。所以 备忘的应该是：

- torch.split
- torch.stack
- torch.randperm
- roi_align（外部）
- nms（外部）

**Comment by simonJJJ** [comment]
> > sigmoid
> 
> 这里的 topk、max、min 之江黄振华正在做。sigmoid 已经有了。zeros_like应该可以已有的用zeros替代。所以 备忘的应该是：
> 
> * torch.split
> * torch.stack
> * torch.randperm
> * roi_align（外部）
> * nms（外部）

topk有pr吗？没有的话我可以来写

**Comment by ScXfjiang** [comment]
> Mask R-CNN所需算子：
> 
> pytorch	oneflow待搬运	备注	认领人
> torch.nn.ConvTranspose2d	flow.nn.ConvTranspose2d		张晓雨
> torch.topk	flow.topk		
> torch.split	flow.split		
> torch.max	flow.max		
> torch.min	flow.min		
> torch.stack	flow.stack		
> torch.randperm	flow.randperm		
> roi_align(外部op)	roi_align		
> nms(外部op)	nms

oneflow还需要COCO Reader Module

**Comment by BBuf** [comment]
MobileNetV3还需要的算子：
